File: api/download.py
"""
Vercel serverless function — generates CSV or Excel download from posted rows.
POST /api/download?format=csv|excel
Body: JSON array of row objects
"""

from http.server import BaseHTTPRequestHandler
from urllib.parse import urlparse, parse_qs
import json
import io

import pandas as pd


class handler(BaseHTTPRequestHandler):
    def do_POST(self):
        parsed = urlparse(self.path)
        qs = parse_qs(parsed.query)
        fmt = qs.get("format", ["csv"])[0]

        length = int(self.headers.get("Content-Length", 0))
        body = self.rfile.read(length)
        rows = json.loads(body)
        df = pd.DataFrame(rows)

        if fmt == "excel":
            buf = io.BytesIO()
            df_unique = df.drop_duplicates(subset=["paa_question"])
            with pd.ExcelWriter(buf, engine="openpyxl") as writer:
                df.to_excel(writer, sheet_name="All PAA", index=False)
                df_unique.to_excel(writer, sheet_name="Unique Questions", index=False)
                for kw in df["seed_keyword"].unique():
                    df[df["seed_keyword"] == kw].to_excel(
                        writer, sheet_name=kw[:31], index=False
                    )
            data = buf.getvalue()
            mime = "application/vnd.openxmlformats-officedocument.spreadsheetml.sheet"
            filename = "paa_results.xlsx"
        else:
            buf = io.StringIO()
            df.to_csv(buf, index=False)
            data = buf.getvalue().encode()
            mime = "text/csv"
            filename = "paa_results.csv"

        self.send_response(200)
        self.send_header("Content-Type", mime)
        self.send_header("Content-Disposition", f'attachment; filename="{filename}"')
        self.send_header("Content-Length", str(len(data)))
        self.send_header("Access-Control-Allow-Origin", "*")
        self.end_headers()
        self.wfile.write(data if isinstance(data, bytes) else data.encode())

    def do_OPTIONS(self):
        self.send_response(200)
        self.send_header("Access-Control-Allow-Origin", "*")
        self.send_header("Access-Control-Allow-Methods", "POST, OPTIONS")
        self.send_header("Access-Control-Allow-Headers", "Content-Type")
        self.end_headers()

    def log_message(self, *args):
        pass

File: api/download.js
/**
 * Vercel serverless function — generates CSV or Excel from posted rows.
 * POST /api/download?format=csv|excel
 * Body: JSON array of row objects
 */

import * as XLSX from "xlsx";

const COLUMNS = [
  "seed_keyword", "depth", "parent_question", "paa_question",
  "snippet", "source_title", "source_link", "displayed_link", "date",
];

function toCSV(rows) {
  const escape = (v) => `"${String(v ?? "").replace(/"/g, '""')}"`;
  const header = COLUMNS.join(",");
  const lines = rows.map((r) => COLUMNS.map((c) => escape(r[c])).join(","));
  return [header, ...lines].join("\n");
}

function toExcel(rows) {
  const wb = XLSX.utils.book_new();

  // All PAA sheet
  const ws1 = XLSX.utils.json_to_sheet(rows, { header: COLUMNS });
  XLSX.utils.book_append_sheet(wb, ws1, "All PAA");

  // Unique questions sheet
  const seen = new Set();
  const unique = rows.filter((r) => {
    if (seen.has(r.paa_question)) return false;
    seen.add(r.paa_question);
    return true;
  });
  const ws2 = XLSX.utils.json_to_sheet(unique, { header: COLUMNS });
  XLSX.utils.book_append_sheet(wb, ws2, "Unique Questions");

  // Per-keyword sheets
  const byKeyword = {};
  for (const r of rows) {
    (byKeyword[r.seed_keyword] = byKeyword[r.seed_keyword] || []).push(r);
  }
  for (const [kw, kwRows] of Object.entries(byKeyword)) {
    const ws = XLSX.utils.json_to_sheet(kwRows, { header: COLUMNS });
    XLSX.utils.book_append_sheet(wb, ws, kw.slice(0, 31));
  }

  return XLSX.write(wb, { type: "buffer", bookType: "xlsx" });
}

export default async function handler(req, res) {
  if (req.method !== "POST") return res.status(405).end();

  const format = req.query.format || "csv";
  const rows = req.body;

  if (!Array.isArray(rows)) return res.status(400).json({ error: "Body must be a JSON array." });

  if (format === "excel") {
    const buf = toExcel(rows);
    res.setHeader("Content-Type", "application/vnd.openxmlformats-officedocument.spreadsheetml.sheet");
    res.setHeader("Content-Disposition", 'attachment; filename="paa_results.xlsx"');
    res.send(buf);
  } else {
    const csv = toCSV(rows);
    res.setHeader("Content-Type", "text/csv");
    res.setHeader("Content-Disposition", 'attachment; filename="paa_results.csv"');
    res.send(csv);
  }
}
